index.js

```diff
@@ -1,43 +1,62 @@
 /* eslint-env node */
 'use strict';
+const Funnel = require('broccoli-funnel');
+const getDebugMacros = require('./src/debug-macros').debugMacros;
 
-function debugMacros(features) {
-  let plugins = [
-    [
-      require.resolve('babel-plugin-debug-macros'),
-      {
-        flags: [
-          {
-            source: 'ember-m3/feature-flags',
-            flags: features,
-          },
-        ],
-      },
-      '@ember-data/canary-features-stripping',
-    ],
-  ];
-
-  return plugins;
+/*
+ shouldIncludeChildAddon runs during
+ addon init so does not have access to
+ the host application instance.
+*/
+function hasDataPackage(addon) {
+  let addons = addon.project.addonPackages;
+  return addons['ember-data'] !== undefined;
 }
 
 module.exports = {
   name: 'ember-m3',
 
-  init() {
-    this._super.init.apply(this, arguments);
+  included() {
+    this._super.included.call(this, ...arguments);
+    this.configureBabelOptions();
+  },
+
+  shouldIncludeChildAddon(addon) {
+    if (!hasDataPackage(this)) {
+      return true;
+    }
+    if (addon.name.startsWith('@ember-data')) {
+      /*
+      console.log(
+        `⚠️  ember-m3 is excluding ${addon.name} version ${addon.pkg.version} from the build`
+      );
+      */
+      return false;
+    }
+    return true;
+  },
+
+  treeForAddon(tree) {
+    const isProd = process.env.EMBER_ENV === 'production';
+    const exclude = [];
 
-    let features;
-    try {
-      features = this.project.require('@ember-data/private-build-infra/src/features')();
-    } catch (e) {
-      features = { CUSTOM_MODEL_CLASS: false };
+    if (isProd) {
+      exclude.push('adapters');
     }
+    tree = new Funnel(tree, {
+      exclude,
+    });
+    return this._super.treeForAddon.call(this, tree);
+  },
+
+  configureBabelOptions() {
+    let app = this._findHost();
 
     this.options = this.options || {};
     this.options.babel = this.options.babel || {};
-    // this ensures that the same `@ember-data/canary-features` processing that the various
-    // ember-data addons do is done in the dummy app
-    this.options.babel.plugins = [...debugMacros(features)];
+    let plugins = this.options.babel.plugins;
+    let newPlugins = getDebugMacros(app);
+    this.options.babel.plugins = Array.isArray(plugins) ? plugins.concat(newPlugins) : newPlugins;
```

Comment:
Inserting these debug macros globally  is surprisingly, do we not want to only run them for the subset that need them?

Reply:
@stefanpenner what do you mean by global? this.options.babel will apply only to the addon, not the consuming app.